File: go/plugins/doppler/doppler.go
// The doppler plugin: Doppler, one bulk download per config. Needs
// HTTPS. A port of typescript/plugins/doppler.ts.
package doppler

import (
	"net/http"
	"strconv"
	"strings"
	"sync"

	"github.com/voxgig/sekreto/go/plugins/httpjson"
	"github.com/voxgig/sekreto/go/sekreto"
)

// Provider reads Doppler.
//
// The whole config is downloaded once - Doppler's own bulk endpoint - and
// answered from memory, like a remote .env: api.token is the API_TOKEN
// entry. A service token is config-scoped, so project and config are only
// needed with broader tokens.
type Provider struct {
	Token   string
	Project string
	Config  string
	Addr    string
	// Guards the memoised state below: a Sekreto may resolve from several
	// goroutines, and a racy map read is either a crash or a zero value - a
	// MISS where the store does hold the secret, which falls through to a
	// weaker store.
	mu     sync.Mutex
	values map[string]string
}

func (provider *Provider) load() (map[string]string, error) {
	provider.mu.Lock()
	defer provider.mu.Unlock()

	if nil != provider.values {
		return provider.values, nil
	}

	addr := provider.Addr
	if "" == addr {
		addr = "https://api.doppler.com"
	}
	addr = strings.TrimSuffix(addr, "/")
	if err := sekreto.CheckAddr(addr); nil != err {
		return nil, err
	}

	target := addr + "/v3/configs/config/secrets/download?format=json"
	if "" != provider.Project {
		target += "&project=" + httpjson.Escape(provider.Project)
	}
	if "" != provider.Config {
		target += "&config=" + httpjson.Escape(provider.Config)
	}

	status, body, err := httpjson.Get(target,
		map[string]string{"authorization": "Bearer " + provider.Token})
	if nil != err {
		return nil, err
	}

	object, is := body.(map[string]any)
	if http.StatusOK != status || !is {
		return nil, sekreto.Fail("sekreto: doppler error: " + strconv.Itoa(status))
	}

	values := map[string]string{}
	for key, value := range object {
		if nil != value {
			values[key] = httpjson.ToString(value)
		}
	}
	provider.values = values

	return values, nil
}

func (provider *Provider) Lookup(name string) (string, bool, error) {
	key, err := sekreto.EnvKey(name, "")
	if nil != err {
		return "", false, err
	}

	values, err := provider.load()
	if nil != err {
		return "", false, err
	}

	value, has := values[key]
	return value, has, nil
}

func (provider *Provider) Describe() string {
	if "" != provider.Project {
		return "doppler:" + provider.Project + "/" + provider.Config
	}
	return "doppler"
}

// Plugin is the `doppler` provider kind, as a voxgig/plugin definition.
var Plugin = sekreto.ProviderPlugin("doppler", func(spec *sekreto.ProviderSpec) (sekreto.Provider, error) {
	return &Provider{
		Token:   spec.Token,
		Project: spec.Project,
		Config:  spec.Config,
		Addr:    spec.Addr,
	}, nil
})

File: lua/src/sekreto/plugins/doppler.lua
-- The doppler plugin: a whole Doppler config downloaded once, over
-- HTTPS, and answered from memory like a remote .env.
--
-- A port of typescript/plugins/doppler.ts, which is canonical.

local err = require('sekreto.err')
local name = require('sekreto.name')
local addr = require('sekreto.addr')
local providers = require('sekreto.providers')
local json = require('sekreto.plugins.json')
local httpjson = require('sekreto.plugins.httpjson')
local support = require('sekreto.plugins.support')

local fail = err.fail
local nonempty = providers.nonempty
local trimslash = providers.trimslash
local providerplugin = providers.providerplugin
local first = support.first
local fetchjson = httpjson.fetchjson
local uriescape = support.uriescape

local M = {}

--- Doppler.
---
--- The whole config is downloaded once - Doppler's own bulk endpoint -
--- and answered from memory, like a remote .env: `api.token` is the
--- `API_TOKEN` entry. A failed load caches nothing, so it retries.
local function doppler(token, project, config, useaddr)
  local values = nil

  local function load()
    if nil ~= values then
      return values
    end

    local address = trimslash(first(useaddr, 'https://api.doppler.com'))
    addr.checkaddr(address)

    local url = address .. '/v3/configs/config/secrets/download?format=json'
    if nonempty(project) then
      url = url .. '&project=' .. uriescape(project)
    end
    if nonempty(config) then
      url = url .. '&config=' .. uriescape(config)
    end

    local res = fetchjson(
      'GET', url, { { 'authorization', 'Bearer ' .. (token or '') } }
    )

    local body = json.asobj(res.body)
    if 200 ~= res.status or nil == body then
      fail('sekreto: doppler error: ' .. res.status)
    end

    local loaded = {}
    for _, key in ipairs(body.keys) do
      local text = json.text(body.vals[key])
      if nil ~= text then
        loaded[key] = text
      end
    end

    values = loaded
    return loaded
  end

  return {
    -- The `prefix` option is deliberately not consulted by this kind.
    lookup = function(secret)
      return load()[name.envkey(secret)]
    end,
    describe = function()
      return 'doppler' ..
        (nonempty(project) and (':' .. project .. '/' .. (config or '')) or '')
    end,
  }
end

--- The kind, as a voxgig/plugin definition.
M.doppler = providerplugin('doppler', function(spec)
  return doppler(spec.token, spec.project, spec.config, spec.addr)
end)

return M
